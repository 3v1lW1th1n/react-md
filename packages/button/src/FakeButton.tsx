import * as React from "react";
import { IWithForwardedRef, omit } from "@react-md/utils";

import { IButtonDefaultProps } from "./Button";
import ButtonChildren from "./ButtonChildren";

import { IButtonThemeProps } from "./types.d";
import theme from "./theme";

export interface IFakeButtonProps
  extends IButtonThemeProps,
    React.HTMLAttributes<HTMLDivElement> {}

export interface IFakeButtonDefaultProps extends IButtonDefaultProps {
  role: string;
  tabIndex: number;
}

export type FakeButtonWithDefaultProps = IFakeButtonProps &
  IFakeButtonDefaultProps &
  IWithForwardedRef<HTMLDivElement>;

/**
 * Need to remove the event listeners when the disabled prop
 * is enabled since there is no "real" disabled state built
 * into divs. Instead we will need to fake it programatically
 * by setting `aria-disabled="true"` and removing the event
 * listeners from the div.
 */
const EVENT_LISTENERS = [
  "onKeyUp",
  "onKeyPress",
  "onMouseDown",
  "onMouseUp",
  "onMouseOver",
  "onMouseEnter",
  "onMouseLeave",
  "onTouchStart",
  "onTouchMove",
  "onTouchEnd",
];

/**
 * The `FakeButton` is used for places where you need an element to interact
 * like a button, but want to have separate clickable children in the button
 * or "advanced" html of using nested `<div>`s since it is _technically_
 * invalid HTML to have `<div>`s as a descendant of a button.
 *
 * This component will render as an accessible "button" div by following the
 * button role specifications of:
 * - apply the button role to the div
 * - applying a tab index of 0 so it is keyboard focusable
 * - attaching a keydown event listener to click and the div when the user
 *   presses space or enter
 * - apply aria-disable="true" when the `disabled` prop has been enabled and
 *   removing all event listeners until it is enabled again.
 *
 * @see https://html.spec.whatwg.org/multipage/form-elements.html#the-button-element
 * @see https://html.spec.whatwg.org/multipage/dom.html#phrasing-content-2
 */
const FakeButton: React.FunctionComponent<
  IFakeButtonProps & IWithForwardedRef<HTMLDivElement>
> = providedProps => {
  const {
    theme: propTheme,
    themeType,
    buttonType,
    children,
    forwardedRef,
    preventColorPollution,
    disabled,
    tabIndex,
    onKeyDown,
    ...props
  } = providedProps as FakeButtonWithDefaultProps;

  function handleKeyDown(event: React.KeyboardEvent<HTMLDivElement>) {
    if (onKeyDown) {
      onKeyDown(event);
    }

    const { key } = event;
    const isSpace = key === " ";
    if (isSpace || key === "Enter") {
      if (isSpace) {
        event.preventDefault();
      }

      event.currentTarget.click();
    }
  }

  return (
    <div
      aria-disabled={disabled ? "true" : undefined}
      {...(disabled ? omit(props, EVENT_LISTENERS) : props)}
      ref={forwardedRef}
      className={theme(providedProps)}
      tabIndex={disabled ? undefined : tabIndex}
      onKeyDown={disabled ? undefined : handleKeyDown}
    >
      <ButtonChildren preventColorPollution={preventColorPollution}>
        {children}
      </ButtonChildren>
    </div>
  );
};

const defaultProps: IFakeButtonDefaultProps = {
  tabIndex: 0,
  role: "button",
  disabled: false,
  theme: "primary",
  themeType: "flat",
  buttonType: "text",
  preventColorPollution: false,
};

FakeButton.defaultProps = defaultProps;

if (process.env.NODE_ENV !== "production") {
  let PropTypes = null;
  try {
    PropTypes = require("prop-types");
  } catch (e) {}

  if (PropTypes) {
    FakeButton.propTypes = {
      tabIndex: PropTypes.number,
      role: PropTypes.string,
      className: PropTypes.string,
      theme: PropTypes.oneOf([
        "clear",
        "primary",
        "secondary",
        "warning",
        "error",
      ]),
      themeType: PropTypes.oneOf(["flat", "outline", "contained"]),
      buttonType: PropTypes.oneOf(["text", "icon"]),
      disabled: PropTypes.bool,
      children: PropTypes.node,
    };
  }
}

export default React.forwardRef<HTMLDivElement, IFakeButtonProps>(
  (props, ref) => <FakeButton {...props} forwardedRef={ref} />
);
