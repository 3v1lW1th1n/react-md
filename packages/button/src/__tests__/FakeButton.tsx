import * as React from "react";
import { mount } from "enzyme";
import { create } from "react-test-renderer";

import FakeButton from "../FakeButton";

import {
  ButtonTheme,
  ButtonThemeType,
  ButtonType,
  IButtonThemeProps,
} from "../types";

const themes: ButtonTheme[] = [
  "clear",
  "primary",
  "secondary",
  "warning",
  "error",
  "default",
];
const themeTypes: ButtonThemeType[] = ["flat", "contained", "outline"];
const buttonTypes: ButtonType[] = ["text", "icon"];

function flattenDeep(
  arr: any[]
): (IButtonThemeProps & { children: React.ReactNode })[] {
  return arr.reduce(
    (flattened, val) =>
      Array.isArray(val)
        ? flattened.concat(flattenDeep(val))
        : flattened.concat(val),
    []
  );
}

describe("FakeButton", () => {
  it("should apply the aria-disabled attribute when the disabled prop is enabled", () => {
    const button = mount(<FakeButton>Button</FakeButton>);
    let div = button.find("div");

    expect(div.props()["aria-disabled"]).toBeUndefined();
    expect(div.props().disabled).toBeUndefined();

    button.setProps({ disabled: true });
    div = button.find("div");

    expect(div.props()["aria-disabled"]).toBe("true");
    expect(div.props().disabled).toBeUndefined();
  });

  it("should remove the tabIndex prop when the disabled prop is enabled", () => {
    const button = mount(<FakeButton>Button</FakeButton>);
    let div = button.find("div");

    expect(div.props().tabIndex).toBe(0);

    button.setProps({ disabled: true });
    div = button.find("div");
    expect(div.props().tabIndex).toBeUndefined();
  });

  it("should correctly call the onKeyDown prop", () => {
    const onKeyDown = jest.fn();
    const button = mount(<FakeButton onKeyDown={onKeyDown}>Button</FakeButton>);

    button.simulate("keyDown");
    expect(onKeyDown).toBeCalled();
  });

  it("should call preventDefault when the space key is pressed to prevent page scrolling", () => {
    const button = mount(<FakeButton>Button</FakeButton>);

    const preventDefault = jest.fn();
    button.simulate("keyDown", { preventDefault, key: " " });

    expect(preventDefault).toBeCalled();

    preventDefault.mockClear();
    button.simulate("keyDown", { preventDefault, key: "Enter" });
    expect(preventDefault).not.toBeCalled();
  });

  // Unfortunately I have to skip this test as the currentTarget doesn't seem to be
  // modifiable at this time and it doesn't actually trigger a click event in testing
  it.skip("should correctly trigger the click event after space or enter is pressed", () => {
    const onClick = jest.fn();
    const button = mount(<FakeButton onClick={onClick}>Button</FakeButton>);

    button.simulate("keyDown", { key: " " });
    expect(onClick).toBeCalled();

    onClick.mockClear();
    button.simulate("keyDown", { key: "Enter" });
    expect(onClick).toBeCalled();
  });

  it("should render correctly based on the theme props", () => {
    flattenDeep(
      buttonTypes.map(buttonType =>
        themes.map(theme =>
          themeTypes.map(themeType => ({
            buttonType,
            theme,
            themeType,
            children: buttonType === "text" ? "Button Text" : <i />,
          }))
        )
      )
    ).forEach(themeProps => {
      expect(create(<FakeButton {...themeProps} />).toJSON()).toMatchSnapshot();
    });
  });
});
