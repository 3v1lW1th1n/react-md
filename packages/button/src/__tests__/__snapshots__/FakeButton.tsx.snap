// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FakeButton should render correctly based on the theme props 1`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 2`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--contained"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 3`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--outline"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 4`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--text-primary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 5`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--contained rmd-button--primary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 6`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--outline rmd-button--text-primary rmd-button--outline-primary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 7`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--text-secondary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 8`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--contained rmd-button--secondary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 9`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--outline rmd-button--text-secondary rmd-button--outline-secondary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 10`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--text-warning"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 11`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--contained rmd-button--warning"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 12`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--outline rmd-button--text-warning rmd-button--outline-warning"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 13`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--text-error"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 14`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--contained rmd-button--error"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 15`] = `
<div
  className="rmd-button rmd-button--text rmd-button--hoverable rmd-button--outline rmd-button--text-error rmd-button--outline-error"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 16`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 17`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--contained"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 18`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--outline"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 19`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--text-primary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 20`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--contained rmd-button--primary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 21`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--outline rmd-button--text-primary rmd-button--outline-primary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 22`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--text-secondary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 23`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--contained rmd-button--secondary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 24`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--outline rmd-button--text-secondary rmd-button--outline-secondary"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 25`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--text-warning"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 26`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--contained rmd-button--warning"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 27`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--outline rmd-button--text-warning rmd-button--outline-warning"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 28`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--text-error"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 29`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--contained rmd-button--error"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;

exports[`FakeButton should render correctly based on the theme props 30`] = `
<div
  className="rmd-button rmd-button--icon rmd-button--hoverable rmd-button--outline rmd-button--text-error rmd-button--outline-error"
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    aria-hidden="true"
    className="rmd-button__content"
  />
  <span
    className="rmd-button__content-mask"
  />
</div>
`;
